framework/helpers/BaseArrayHelper.php

```diff
@@ -923,41 +923,51 @@ public static function isSubset($needles, $haystack, $strict = false)
     public static function filter($array, $filters)
     {
         $result = [];
-        $forbiddenVars = [];
-
-        foreach ($filters as $var) {
-            $keys = explode('.', $var);
-            $globalKey = $keys[0];
-            $localKey = isset($keys[1]) ? $keys[1] : null;
-
-            if ($globalKey[0] === '!') {
-                $forbiddenVars[] = [
-                    substr($globalKey, 1),
-                    $localKey,
-                ];
-                continue;
-            }
+        $excludeFilters = [];
 
-            if (!array_key_exists($globalKey, $array)) {
-                continue;
-            }
-            if ($localKey === null) {
-                $result[$globalKey] = $array[$globalKey];
+        foreach ($filters as $filter) {
+            if ($filter[0] === '!') {
+                $excludeFilters[] = substr($filter, 1);
                 continue;
             }
-            if (!isset($array[$globalKey][$localKey])) {
-                continue;
+
+            $nodeValue = $array; //set $array as root node
+            $keys = explode('.', $filter);
+            foreach ($keys as $key) {
+                if (!array_key_exists($key, $nodeValue)) {
+                    continue 2; //Jump to next filter
+                }
+                $nodeValue = $nodeValue[$key];
             }
-            if (!array_key_exists($globalKey, $result)) {
-                $result[$globalKey] = [];
+
+            //We've found a value now let's insert it
+            $resultNode = &$result;
+            foreach ($keys as $key) {
```

Comment:
Can you merge this loop with previous?

Reply:
Nope, let's say we have a filter 'A.B' and the source does have key A but that level does not have key B in it, the resulting array would be set `[A => []]` while it should return `[]`.

Reply:
Are tests fix this case?

Reply:
Yes, the test would fail at line 1336 `$this->assertEquals(ArrayHelper::filter($array, ['A.X']), []);`.